### PR TITLE
trie: add ChildTrie trait to simplify future child trie types addition

### PR DESCRIPTION
**Pull request by sorpaas**
This refactoring adds `ChildTrie` trait. The aim is to simplify child trie types addition.

With this PR, to add a new child trie type, one only needs to:

* Create a new file under `core/trie/src/child_tries`, adds a new unit struct, and implement `ChildTrie`.
* Modify the `with_child_trie` macro, add a new if condition defining the new prefix.

### CODE REVIEW

**Comment by thiolliere** [comment]
tests fail

File: core/trie/src/child_tries/default.rs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+//! the same structure of root trie.
+
+use super::ChildTrie;
+use crate::{TrieHash, TrieError, TrieDBMut, TrieDB};
+use rstd::boxed::Box;
+use rstd::vec::Vec;
+use trie_db::{TrieConfiguration, DBValue, Query, Trie, TrieMut};
+use hash_db::{HashDB, HashDBRef, PlainDB, PlainDBRef};
+
+/// Default child trie.
```
Comment (gavofyork) [inline core/trie/src/child_tries/default.rs]: Useless docs.

File: core/trie/src/child_tries/mod.rs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+pub use self::default::DefaultChildTrie;
+
+use crate::{TrieHash, TrieError};
+use rstd::boxed::Box;
+use rstd::vec::Vec;
+use trie_db::{TrieConfiguration, DBValue, Query};
+use hash_db::{HashDB, HashDBRef, PlainDB, PlainDBRef};
+
+/// Definition for a child trie.
```
Comment (gavofyork) [inline core/trie/src/child_tries/mod.rs]: Useless docs.

File: core/trie/src/child_tries/mod.rs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+pub trait ChildTrie {
+	/// Default root of the child trie.
```
Comment (gavofyork) [inline core/trie/src/child_tries/mod.rs]: Useless docs

**Comment by gnunicorn** [comment]
is still being worked on?

**Comment by sorpaas** [comment]
Sorry. Let me close this for now.

**Pull request closed without merging**